File: code/algo/getoffer/c++/3/genstr_quote_n.cpp
#include <string>
#include <vector>

using namespace std;

vector<string> generateParenthesis(int n) {
  vector<vector<string>> dp(n + 1);
  dp[0] = {""};
  for (int i = 1; i <= n; ++i) {
    for (int j = 0; j < i; ++j) {
      for (const string &left : dp[j]) {
        for (const string &right : dp[i - j - 1]) {
          dp[i].push_back("(" + left + ")" + right);
        }
      }
    }
  }
  return dp[n];
}

void dfs(int n, int leftCnt, int rightCnt, string &tmp, vector<string> &res) {
  if (tmp.size() == n * 2) {
    res.push_back(tmp);
    return;
  }
  if (leftCnt < n) {
    tmp.push_back('(');
    dfs(n, leftCnt + 1, rightCnt, tmp, res);
    tmp.pop_back();
  }
  if (leftCnt > rightCnt) {
    tmp.push_back(')');
    dfs(n, leftCnt, rightCnt + 1, tmp, res);
    tmp.pop_back();
  }
}

vector<string> generateParenthesisDFS(int n) {
  vector<string> res;
  string tmp;
  dfs(n, 0, 0, tmp, res);
  return res;
}

File: code/algo/getoffer/go/3/genstr_quote_n.go
package main

// 动态规划版本
func generateParenthesisDP(n int) []string {
	dp := make([][]string, n+1)
	dp[0] = []string{""}
	for i := 1; i <= n; i++ {
		for j := 0; j < i; j++ {
			for _, left := range dp[j] {
				for _, right := range dp[i-j-1] {
					dp[i] = append(dp[i], "("+left+")"+right)
				}
			}

		}
	}
	return dp[n]
}

// 递归版本
func generateParenthesisDFS(n int) []string {
	var res []string
	var leftCnt, rightCnt int
	var tmp []byte
	var dfs func(cnt int)
	dfs = func(cnt int) {
		if cnt == n*2 {
			res = append(res, string(tmp))
			return
		}
		if leftCnt < n {
			leftCnt++
			tmp = append(tmp, '(')
			dfs(cnt + 1)
			leftCnt--
			tmp = tmp[:len(tmp)-1]
		}
		if leftCnt > rightCnt {
			rightCnt++
			tmp = append(tmp, ')')
			dfs(cnt + 1)
			rightCnt--
			tmp = tmp[:len(tmp)-1]
		}
	}
	dfs(0)
	return res
}
